runtime/Dart/lib/src/tree/src/tree.dart

```diff
@@ -44,15 +44,18 @@ abstract class SyntaxTree extends Tree {
   /// <p>As a weird special case, the source interval for rules matched after
   /// EOF is unspecified.</p>
   Interval get sourceInterval;
+
+  @override
+  T? getChild<T>(int i);
 }
 
 abstract class ParseTree extends SyntaxTree {
   // the following methods narrow the return type; they are not additional methods
   @override
-  ParseTree get parent;
+  ParseTree? get parent;
 
   @override
-  ParseTree getChild<T>(int i);
+  T? getChild<T>(int i);
```

Comment:
This will return dynamic when used comparing to before. What is the reason?

Reply:
That was probably due to some incompatibility issue that we found on the process of migranting to NNBD, gonna rollback some spcefics regarding this now.

Reply:
Ok now I got to remember why: since 2.12, things like [this](https://github.com/antlr/antlr4/blob/2cb1618009049f751b83e1538ecc659da6a7e677/runtime/Dart/lib/src/parser_rule_context.dart#L203) wont be possible anymore: a method that returns as supertype be automatically casted to a supertype, in this case,  `getChild` returns `ParseTree` that is auto casted to the generic type, subclasses of `ParserRuleContext`.

Reply:
At first we tried to pass the generic as a subclass of tree (or parsetree in the subclass: `<T extends ParseTree>`)but dart doesn't let overwrites as such. Rolling back to manual overwrites since it would be a less disrupt change.

Reply:
I'd suggest documenting this change for clarity.

Reply:
It was changed back to ParseTree @jnhyatt